apm-agent-plugins/apm-profiling-plugin/src/main/java/co/elastic/apm/agent/profiler/CallTree.java

```diff
@@ -0,0 +1,422 @@
+ * aaaa ->  ├─b    1
+ *          └─b    1
+ * </pre>
+ * <p>
+ * It also stores information about which span is the parent of a particular call tree node,
+ * based on which span has been {@linkplain ElasticApmTracer#getActive() active} at that time.
+ * </p>
+ * <p>
+ * This allows to {@linkplain Root#spanify() infer spans from the call tree} which have the correct parent/child relationships
+ * with the regular spans.
+ * </p>
+ */
+public class CallTree implements Recyclable {
+
+    @Nullable
+    private CallTree parent;
+    protected int count;
+    private List<CallTree> children = new ArrayList<>();
+    @Nullable
+    private StackFrame frame;
+    protected long start;
+    private long lastSeen;
+    private boolean ended;
+    private long activationTimestamp = -1;
+    @Nullable
+    private TraceContext activeContext;
+    private long deactivationTimestamp = -1;
+    private boolean isSpan;
+
+    public CallTree() {
+    }
+
+    public void set(@Nullable CallTree parent, StackFrame frame, long nanoTime) {
+        this.parent = parent;
+        this.frame = frame;
+        this.start = nanoTime;
+    }
+
+    public void activation(TraceContext traceContext, long activationTimestamp) {
+        this.activeContext = traceContext;
+        this.activationTimestamp = activationTimestamp;
+    }
+
+    protected void handleDeactivation(TraceContext deactivatedSpan, long timestamp) {
+        if (deactivatedSpan.equals(activeContext)) {
+            deactivationTimestamp = timestamp;
+        } else {
+            CallTree lastChild = getLastChild();
+            if (lastChild != null && !lastChild.isEnded()) {
+                lastChild.handleDeactivation(deactivatedSpan, timestamp);
+            }
+        }
+    }
+
+    public static CallTree.Root createRoot(ObjectPool<CallTree.Root> rootPool, byte[] traceContext, @Nullable String serviceName, long nanoTime) {
+        CallTree.Root root = rootPool.createInstance();
+        root.set(traceContext, serviceName, nanoTime);
+        return root;
+    }
+
+    protected void addFrame(ListIterator<StackFrame> iterator, @Nullable TraceContext traceContext, long activationTimestamp, long nanoTime) {
+        count++;
+        lastSeen = nanoTime;
+        //     c ee   <- traceContext not set - they are not a child of the active span but the frame below them
+        //   bbb dd   <- traceContext set
+        //   ------   <- all new CallTree during this period should have the traceContext set
+        // a aaaaaa a
+        //  |      |
+        // active  deactive
+
+        // this branch is already aware of the activation
+        if (Objects.equals(this.activeContext, traceContext)) {
+            traceContext = null;
+        }
+
+        CallTree lastChild = getLastChild();
+        // if the frame corresponding to the last child is not in the stack trace
+        // it's assumed to have ended one tick ago
+        boolean endChild = true;
+        if (iterator.hasPrevious()) {
```

Comment:
If you find an easy and cheap way (i.e. without getting the full stack trace) to get Transaction's frame in the root, you can cut all the unnecessary call tree "trunk".

Reply:
Right, this is not for now. Would make sense within this PR only if it is simple enough to do without overhead. We can consider as future enhancement if the need arises.